### PR TITLE
Migrated `pkg/controller/replicaset` to contextual logging

### PR DESCRIPTION
**Pull request by Namanl2001**
Signed-off-by: Naman <namanlakhwani@gmail.com>

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Use contextual logging for `pkg/controller/replicaset`
#### Which issue(s) this PR fixes:
Part of https://github.com/kubernetes/enhancements/issues/3077

#### Does this PR introduce a user-facing change?

```release-note
Migrated the replicaset controller (within `kube-controller-manager`) to use [contextual logging](https://k8s.io/docs/concepts/cluster-administration/system-logs/#contextual-logging).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging
```

/wg structured-logging
/area logging
/assign @pohly

### CODE REVIEW

**Comment by Namanl2001** [comment]
remaining non-contextual logging in the `replica_set.go` requires changes in `staging`, therefore, skipping those in this PR

/remove-sig api-machinery
/remove-sig apps
/remove-sig testing

/sig instrumentation

File: pkg/controller/replicaset/replica_set.go [hunk 1]
```diff
@@ -593,7 +593,7 @@ func (rsc *ReplicaSetController) manageReplicas(ctx context.Context, filteredPod
 		// The skipped pods will be retried later. The next controller resync will
 		// retry the slow start process.
 		if skippedPods := diff - successfulCreations; skippedPods > 0 {
-			klog.V(2).Infof("Slow-start failure. Skipping creation of %d pods, decrementing expectations for %v %v/%v", skippedPods, rsc.Kind, rs.Namespace, rs.Name)
+			klog.FromContext(ctx).V(2).Info("Slow-start failure. Skipping creation of pods, decrementing expectations", "pods skipped", skippedPods, "kind", rsc.Kind, "namespace", rs.Namespace, "name", rs.Name)
```
Comment (pohly) [inline pkg/controller/replicaset/replica_set.go:596]:
```suggestion
			klog.FromContext(ctx).V(2).Info("Slow-start failure. Skipping creation of pods, decrementing expectations", "podsSkipped", skippedPods, "kind", rsc.Kind, "replicaSet", klog.KObj(rs))
```

File: pkg/controller/replicaset/replica_set.go [hunk 2]
```diff
@@ -793,14 +793,14 @@ func (rsc *ReplicaSetController) getIndirectlyRelatedPods(rs *apps.ReplicaSet) (
 		}
 		for _, pod := range pods {
 			if otherRS, found := seen[pod.UID]; found {
-				klog.V(5).Infof("Pod %s/%s is owned by both %v %s/%s and %v %s/%s", pod.Namespace, pod.Name, rsc.Kind, otherRS.Namespace, otherRS.Name, rsc.Kind, relatedRS.Namespace, relatedRS.Name)
+				logger.V(5).Info("Pod is owned by both", "pod namespace", pod.Namespace, "pod name", pod.Name, "kind", rsc.Kind, "namespace", otherRS.Namespace, "name", otherRS.Name, "kind", rsc.Kind, "namespace", relatedRS.Namespace, "name", relatedRS.Name)
```
Comment (pohly) [inline pkg/controller/replicaset/replica_set.go:796]:
```suggestion
				logger.V(5).Info("Pod is owned by both", "pod", klog.KObj(pod), "kind", rsc.Kind, "replicaSets", klog.KObjSlice([]klog.KMetadata{otherRS, relatedRS}))
```

File: pkg/controller/replicaset/replica_set.go [hunk 3]
```diff
@@ -793,14 +793,14 @@ func (rsc *ReplicaSetController) getIndirectlyRelatedPods(rs *apps.ReplicaSet) (
...
 				continue
 			}
 			seen[pod.UID] = relatedRS
 			relatedPods = append(relatedPods, pod)
 		}
 	}
-	klog.V(4).InfoS("Found related pods", "kind", rsc.Kind, "object", klog.KObj(rs), "pods", klog.KObjSlice(relatedPods))
+	logger.V(4).Info("Found related pods", "kind", rsc.Kind, "object", klog.KObj(rs), "pods", klog.KObjSlice(relatedPods))
```
Comment (pohly) [inline pkg/controller/replicaset/replica_set.go:803]:
```suggestion
	logger.V(4).Info("Found related pods", "kind", rsc.Kind, "replicaSet", klog.KObj(rs), "pods", klog.KObjSlice(relatedPods))
```

File: pkg/controller/replicaset/replica_set_test.go [hunk 4]
```diff
@@ -444,6 +447,7 @@ func BenchmarkGetReplicaSetsWithSameController(b *testing.B) {
 	stopCh := make(chan struct{})
 	defer close(stopCh)
 	controller, informers := testNewReplicaSetControllerFromClient(clientset.NewForConfigOrDie(&restclient.Config{Host: "", ContentConfig: restclient.ContentConfig{GroupVersion: &schema.GroupVersion{Group: "", Version: "v1"}}}), stopCh, BurstReplicas)
+	_, ctx := ktesting.NewTestContext(b)
```
Comment (pohly) [inline pkg/controller/replicaset/replica_set_test.go:450]:
```suggestion
	logger, _ := ktesting.NewTestContext(b)
```

File: pkg/controller/replicaset/replica_set_test.go [hunk 5]
```diff
@@ -424,7 +427,7 @@ func TestGetReplicaSetsWithSameController(t *testing.T) {
 		for _, r := range c.rss {
 			informers.Apps().V1().ReplicaSets().Informer().GetIndexer().Add(r)
 		}
-		actualRSs := manager.getReplicaSetsWithSameController(c.rs)
+		actualRSs := manager.getReplicaSetsWithSameController(klog.FromContext(ctx), c.rs)
```
Comment (pohly) [inline pkg/controller/replicaset/replica_set_test.go:430]:
```suggestion
		actualRSs := manager.getReplicaSetsWithSameController(logger, c.rs)
```

File: pkg/controller/replicaset/replica_set_test.go [hunk 6]
```diff
@@ -64,6 +65,7 @@ func testNewReplicaSetControllerFromClient(client clientset.Interface, stopCh ch
 	informers := informers.NewSharedInformerFactory(client, controller.NoResyncPeriodFunc())
 
 	ret := NewReplicaSetController(
+		klog.FromContext(context.TODO()),
```
Comment (pohly) [inline pkg/controller/replicaset/replica_set_test.go:68]: Please use ktesting.

File: pkg/controller/replicaset/replica_set_test.go [hunk 7]
```diff
@@ -391,6 +393,7 @@ func TestGetReplicaSetsWithSameController(t *testing.T) {
 	pendingDeletionRS.ObjectMeta.OwnerReferences[0].UID = "789"
 	now := metav1.Now()
 	pendingDeletionRS.DeletionTimestamp = &now
+	_, ctx := ktesting.NewTestContext(t)
```
Comment (pohly) [inline pkg/controller/replicaset/replica_set_test.go:396]:
```suggestion
	logger, _ := ktesting.NewTestContext(t)
```

File: pkg/controller/replicaset/replica_set_test.go [hunk 8]
```diff
@@ -463,7 +467,7 @@ func BenchmarkGetReplicaSetsWithSameController(b *testing.B) {
 	b.ReportAllocs()
 	b.ResetTimer()
 	for n := 0; n < b.N; n++ {
-		gotRSs := controller.getReplicaSetsWithSameController(targetRS)
+		gotRSs := controller.getReplicaSetsWithSameController(klog.FromContext(ctx), targetRS)
```
Comment (pohly) [inline pkg/controller/replicaset/replica_set_test.go:470]:
```suggestion
		gotRSs := controller.getReplicaSetsWithSameController(logger, targetRS)
```

Comment (pohly) [reply pkg/controller/replicaset/replica_set_test.go:470]: I'll stop pointing this out, please make the change also in the other tests.

File: pkg/controller/replicaset/replica_set_test.go [hunk 9]
```diff
@@ -1240,7 +1250,7 @@ func TestExpectationsOnRecreate(t *testing.T) {
 	}
 
 	err = wait.PollImmediate(100*time.Millisecond, informerSyncTimeout, func() (bool, error) {
-		klog.V(8).Infof("Waiting for queue to have 1 item, currently has: %d", manager.queue.Len())
+		logger.V(8).Info("Waiting for queue to have 1 item", "currently has", manager.queue.Len())
```
Comment (pohly) [inline pkg/controller/replicaset/replica_set_test.go:1253]:
```suggestion
		logger.V(8).Info("Waiting for queue to have 1 item", "length", manager.queue.Len())
```

Comment (pohly) [reply pkg/controller/replicaset/replica_set_test.go:1253]: Also below.

**Review by pohly — CHANGES_REQUESTED** [review]
Several keys probably wouldn't have passed the new, stricter logcheck test for naming of keys.

File: pkg/controller/replicaset/replica_set_test.go (hunk 6 above)
Comment (Namanl2001) [reply pkg/controller/replicaset/replica_set_test.go:68]: @pohly `ktesting.NewTestContext(t)` requires a testing parameter to be passed into it but in this case we don't have one, so how to deal with this?

PS: I have addressed all other comments locally, and will push once the above thing is cleared

**Comment by pohly** [comment]
@Namanl2001 : please check and fix the errors reported in https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/114871/pull-kubernetes-verify/1625816367603126272

File: pkg/controller/replicaset/replica_set_test.go (hunk 6 above)
Comment (pohly) [reply pkg/controller/replicaset/replica_set_test.go:68]: You can either change `testNewReplicaSetControllerFromClient` to accept a `t` parameter or add a `ctx`.

Which one is better depends on how it is used: if it's the first thing that all tests do *and* they don't need a logger or ctx themselves, then doing the `NewTestContext` call here may be most suitable. The `t` parameter may also be useful for registering cleanup functions.

But if the tests also need a ctx or logger, then it's better to initialize those in the tests and pass a ctx here.

Comment (Namanl2001) [reply pkg/controller/replicaset/replica_set_test.go:68]: > if it's the first thing that all tests do and they don't need a logger or ctx themselves

the above is the case with most unit tests so changed it to `testNewReplicaSetControllerFromClient(t *testing.T,...)` and then doing this `logger, _ := ktesting.NewTestContext(t)` inside

however few challenges, I'm facing:
1. the parent func of [this call](https://github.com/kubernetes/kubernetes/blob/5303bb1832cab43bccfa41159a7d1fdcaf6ffa81/pkg/controller/replicaset/replica_set_test.go#L1483) doesn't have `t` parameter
2. the parent func of [this call](https://github.com/kubernetes/kubernetes/blob/5303bb1832cab43bccfa41159a7d1fdcaf6ffa81/pkg/controller/replicaset/replica_set_test.go#L449) has benchmark `b` as input which can't be passed 

que: `The t parameter may also be useful for registering cleanup functions.` is cleanup automatic or do we need to add something for this too?

File: pkg/controller/replicaset/replica_set.go [hunk 10]
```diff
@@ -631,7 +631,7 @@ func (rsc *ReplicaSetController) manageReplicas(ctx context.Context, filteredPod
 					podKey := controller.PodKey(targetPod)
 					rsc.expectations.DeletionObserved(rsKey, podKey)
 					if !apierrors.IsNotFound(err) {
-						klog.V(2).Infof("Failed to delete %v, decremented expectations for %v %s/%s", podKey, rsc.Kind, rs.Namespace, rs.Name)
+						klog.FromContext(ctx).V(2).Info("Failed to delete pod, decremented expectations", "pod", podKey, "kind", rsc.Kind, "namespace", rs.Namespace, "name", rs.Name)
```
Comment (pohly) [inline pkg/controller/replicaset/replica_set.go:634]: Use KObj here, please.

File: pkg/controller/replicaset/replica_set_test.go (hunk 6 above)
Comment (pohly) [reply pkg/controller/replicaset/replica_set_test.go:68]: - change `setupManagerWithGCEnabled` to take a context and use that to pass down logger and cancelation (`stopCh` -> `context.Done()`, with WithCancel in the parent)
- to deal with functions which get used in both tests and benchmarks, you can accept `tb testing.TB`

File: pkg/controller/replicaset/replica_set_test.go (hunk 8 above)
Comment (pohly) [inline pkg/controller/replicaset/replica_set_test.go:470]: Don't call `klog.FromContext(ctx)` inside the benchmarks' for loop. Instead, get a `logger` from `ktesting.NewTestContext(b)`.

**Review by pohly — CHANGES_REQUESTED** [review]

File: pkg/controller/replicaset/replica_set_test.go (hunk 6 above)
Comment (Namanl2001) [reply pkg/controller/replicaset/replica_set_test.go]: - > change setupManagerWithGCEnabled to take a context

I changed it to take `t` from tests

**Review by pohly — COMMENTED** [review]
/lgtm
/label tide/merge-method-squash
/assign @soltysh 

For approval.

**Review by soltysh — APPROVED** [review]
/approve

**Pull request merged**